### PR TITLE
Fix/panic in `GETRANGE`

### LINKED ISSUE #2095: Panic on get range
It seems the panic leads to node restart
```
2022-11-24T06:47:13.427Z	debug	get/remote.go:14	processing node...	{"component": "Object.Get service", "request": "GET_RANGE", "address": "HzfGZcfbknxYZNsUtTg6uuwfogfNdZSZnSSRRerzucjs/9pEeWehubNQfWkSxgYA25mooUQ5q5dLoTMxoxFC4uXAL", "raw": false, "local": false, "with session": false, "with bearer": false}
panic: zero range length

goroutine 28297 [running]:
github.com/nspcc-dev/neofs-sdk-go/client.(*Client).ObjectRangeInit(0xc0002004d0, {0x1266fd8, 0xc0007c9aa0}, {{{0x0, 0x1, 0x0, {0x0, 0x0, 0x0}, 0x0, ...}, ...}, ...})
	github.com/nspcc-dev/neofs-sdk-go@v1.0.0-rc.7/client/object_get.go:678 +0x39d
github.com/nspcc-dev/neofs-node/pkg/network/cache.(*multiClient).ObjectRangeInit.func1({0x12831d8, 0xc0002004d0})
	github.com/nspcc-dev/neofs-node/pkg/network/cache/multi.go:161 +0x82
github.com/nspcc-dev/neofs-node/pkg/network/cache.(*multiClient).iterateClients.func1({{0x1285560, 0xc000795e30}})
	github.com/nspcc-dev/neofs-node/pkg/network/cache/multi.go:108 +0xc9
github.com/nspcc-dev/neofs-node/pkg/network.AddressGroup.IterateAddresses({0xc000a44700, 0x2, 0x0}, 0xc000708df8)
	github.com/nspcc-dev/neofs-node/pkg/network/group.go:37 +0x62
github.com/nspcc-dev/neofs-node/pkg/network/cache.(*multiClient).iterateClients(0x203000, {0x1266fd8, 0xc0007c9aa0}, 0xc000784268)
	github.com/nspcc-dev/neofs-node/pkg/network/cache/multi.go:96 +0x87
github.com/nspcc-dev/neofs-node/pkg/network/cache.(*multiClient).ObjectRangeInit(0x40d6b4, {0x1266fd8, 0xc0007c9aa0}, {{{0x0, 0x1, 0x0, {0x0, 0x0, 0x0}, 0x0, ...}, ...}, ...})
	github.com/nspcc-dev/neofs-node/pkg/network/cache/multi.go:160 +0xd7
github.com/nspcc-dev/neofs-node/pkg/services/object/internal/client.PayloadRange({{{{0x1283958, 0xc00031d580}, {0x1266fd8, 0xc0007c9aa0}, 0xc0003c0060, 0x0, 0x0, 0x1, {0x1b66af8, 0x0, ...}}}, ...})
	github.com/nspcc-dev/neofs-node/pkg/services/object/internal/client/client.go:354 +0x295
github.com/nspcc-dev/neofs-node/pkg/services/object/get.(*clientWrapper).getObject(0xc001abfc30, 0xc0007a0dd0, {{0xc000a44700, 0x2, 0x2}, {0x0, 0x0, 0x0}, {0xc0005fc510, 0x21, ...}})
	github.com/nspcc-dev/neofs-node/pkg/services/object/get/util.go:143 +0x318
github.com/nspcc-dev/neofs-node/pkg/services/object/get.(*execCtx).processNode(0xc0007a0dd0, {0xc00070a178, 0x938bcf}, {{0xc000a44700, 0x2, 0x2}, {0x0, 0x0, 0x0}, {0xc0005fc510, ...}})
	github.com/nspcc-dev/neofs-node/pkg/services/object/get/remote.go:21 +0x13e
github.com/nspcc-dev/neofs-node/pkg/services/object/get.(*execCtx).processCurrentEpoch(0xc0007a0dd0)
	github.com/nspcc-dev/neofs-node/pkg/services/object/get/container.go:86 +0x3d5
github.com/nspcc-dev/neofs-node/pkg/services/object/get.(*execCtx).executeOnContainer(0xc0007a0dd0)
	github.com/nspcc-dev/neofs-node/pkg/services/object/get/container.go:29 +0x1d5
github.com/nspcc-dev/neofs-node/pkg/services/object/get.(*execCtx).analyzeStatus(0xc0007a0dd0, 0x1)
	github.com/nspcc-dev/neofs-node/pkg/services/object/get/get.go:113 +0x1ac
github.com/nspcc-dev/neofs-node/pkg/services/object/get.(*execCtx).execute(0xc0007a0dd0)
	github.com/nspcc-dev/neofs-node/pkg/services/object/get/get.go:92 +0x57
github.com/nspcc-dev/neofs-node/pkg/services/object/get.(*Service).get(0xc0014aa7b8, {0x1266fd8, 0xc0007c9aa0}, {{0x12597e0, 0xc0009dc180}, 0xc0007cbf40, {{0xfc, 0x80, 0x5d, 0xf7, ...}, ...}, ...}, ...)
	github.com/nspcc-dev/neofs-node/pkg/services/object/get/get.go:81 +0x17e
github.com/nspcc-dev/neofs-node/pkg/services/object/get.(*execCtx).getChild(0xc002337860, {0x82, 0xf7, 0x39, 0x4e, 0xf5, 0xb3, 0xee, 0xbf, 0xbe, ...}, ...)
	github.com/nspcc-dev/neofs-node/pkg/services/object/get/exec.go:207 +0x26b
github.com/nspcc-dev/neofs-node/pkg/services/object/get.(*execCtx).overtakePayloadDirectly(0xc002337860, {0xc000b48800, 0x2, 0x0}, {0xc0000c23a0, 0x2, 0x8}, 0x0)
	github.com/nspcc-dev/neofs-node/pkg/services/object/get/assemble.go:157 +0x12d
github.com/nspcc-dev/neofs-node/pkg/services/object/get.(*execCtx).overtakePayloadInReverse(0xc002337860, {0x66, 0xc8, 0x44, 0x58, 0xdf, 0x4f, 0x86, 0x5c, 0xe6, ...})
	github.com/nspcc-dev/neofs-node/pkg/services/object/get/assemble.go:188 +0x167
github.com/nspcc-dev/neofs-node/pkg/services/object/get.(*execCtx).assemble(0xc002337860)
	github.com/nspcc-dev/neofs-node/pkg/services/object/get/assemble.go:59 +0x2ca
github.com/nspcc-dev/neofs-node/pkg/services/object/get.(*execCtx).analyzeStatus(0xc002337860, 0x1)
	github.com/nspcc-dev/neofs-node/pkg/services/object/get/get.go:104 +0x214
github.com/nspcc-dev/neofs-node/pkg/services/object/get.(*execCtx).execute(0xc002337860)
	github.com/nspcc-dev/neofs-node/pkg/services/object/get/get.go:92 +0x57
github.com/nspcc-dev/neofs-node/pkg/services/object/get.(*Service).get(0xc0014aa7b8, {0x1266fd8, 0xc0007c9aa0}, {{0x1259808, 0xc0007c9e90}, 0xc0007cbf40, {{0xfc, 0x80, 0x5d, 0xf7, ...}, ...}, ...}, ...)
	github.com/nspcc-dev/neofs-node/pkg/services/object/get/get.go:81 +0x17e
github.com/nspcc-dev/neofs-node/pkg/services/object/get.(*Service).getRange(0xc0019d09a0, {0x1266fd8, 0xc0007c9aa0}, {{{0x1259808, 0xc0007c9e90}, 0xc0007cbf40, {{0xfc, 0x80, 0x5d, 0xf7, ...}, ...}, ...}, ...}, ...)
	github.com/nspcc-dev/neofs-node/pkg/services/object/get/get.go:22 +0x179
github.com/nspcc-dev/neofs-node/pkg/services/object/get.(*Service).GetRangeHash(0x11c8c8ccc, {0x1266fd8, 0xc0007c9aa0}, {{{0x0, 0x0}, 0xc0007cbf40, {{0xfc, 0x80, 0x5d, 0xf7, ...}, ...}, ...}, ...})
	github.com/nspcc-dev/neofs-node/pkg/services/object/get/get.go:45 +0x437
github.com/nspcc-dev/neofs-node/pkg/services/object/get/v2.(*Service).GetRangeHash(0xc0014aa7c0, {0x1266fd8, 0xc0007c9aa0}, 0xc0006d52a8)
	github.com/nspcc-dev/neofs-node/pkg/services/object/get/v2/service.go:86 +0xcb
main.(*objectSvc).GetRangeHash(0x21, {0x1266fd8, 0xc0007c9aa0}, 0xc0006d52a8)
	github.com/nspcc-dev/neofs-node/cmd/neofs-node/object.go:97 +0x27
github.com/nspcc-dev/neofs-node/pkg/services/object.TransportSplitter.GetRangeHash(...)
	github.com/nspcc-dev/neofs-node/pkg/services/object/transport_splitter.go:148
github.com/nspcc-dev/neofs-node/pkg/services/object/acl/v2.Service.GetRangeHash({0xc0019c34a0, {0xc00039a0d0, {0x1248720, 0xc0019d09a0}, {0x125c200, 0xc0014aa558}}}, {0x1266fd8, 0xc0007c9aa0}, 0xc0006d52a8)
	github.com/nspcc-dev/neofs-node/pkg/services/object/acl/v2/service.go:442 +0x43e
github.com/nspcc-dev/neofs-node/pkg/services/object.(*Common).GetRangeHash(0xc0019d09c0, {0x1266fd8, 0xc0007c9aa0}, 0x125c140)
	github.com/nspcc-dev/neofs-node/pkg/services/object/common.go:90 +0x8b
github.com/nspcc-dev/neofs-node/pkg/services/object.(*ResponseService).GetRangeHash.func1({0x1266fd8, 0xc0007c9aa0}, {0x1049880, 0xc0006d52a8})
	github.com/nspcc-dev/neofs-node/pkg/services/object/response.go:148 +0x4f
github.com/nspcc-dev/neofs-node/pkg/services/util/response.(*Service).HandleUnaryRequest(0xc00000e110, {0x1266fd8, 0xc0007c9aa0}, {0x1049880, 0xc0006d52a8}, 0x19b)
	github.com/nspcc-dev/neofs-node/pkg/services/util/response/unary.go:13 +0x46
github.com/nspcc-dev/neofs-node/pkg/services/object.(*ResponseService).GetRangeHash(0xc0007cbe50, {0x1266fd8, 0xc0007c9aa0}, 0xc0006d52a8)
	github.com/nspcc-dev/neofs-node/pkg/services/object/response.go:146 +0x50
github.com/nspcc-dev/neofs-node/pkg/services/object.(*SignService).GetRangeHash.func1({0x1266fd8, 0xc0007c9aa0}, {0x1049880, 0xc0006d52a8})
	github.com/nspcc-dev/neofs-node/pkg/services/object/sign.go:198 +0x4f
github.com/nspcc-dev/neofs-node/pkg/services/util.(*SignService).HandleUnaryRequest(0xc0014aa7e0, {0x1266fd8, 0xc0007c9aa0}, {0x1049880, 0xc0006d52a8}, 0xc000c959e0, 0x1126518)
	github.com/nspcc-dev/neofs-node/pkg/services/util/sign.go:194 +0x135
github.com/nspcc-dev/neofs-node/pkg/services/object.(*SignService).GetRangeHash(0x7fa1501e4901, {0x1266fd8, 0xc0007c9aa0}, 0x7fa1770dc108)
	github.com/nspcc-dev/neofs-node/pkg/services/object/sign.go:196 +0x58
github.com/nspcc-dev/neofs-node/pkg/services/object.MetricCollector.GetRangeHash({{0x1278300, 0xc0019d09e0}, {0x1288950, 0xc00000a1e0}}, {0x1266fd8, 0xc0007c9aa0}, 0x1)
	github.com/nspcc-dev/neofs-node/pkg/services/object/metrics.go:137 +0x142
github.com/nspcc-dev/neofs-node/pkg/network/transport/object/grpc.(*Server).GetRangeHash(0xc0019da040, {0x1266fd8, 0xc0007c9aa0}, 0xc000b42ba0)
	github.com/nspcc-dev/neofs-node/pkg/network/transport/object/grpc/service.go:106 +0x74
github.com/nspcc-dev/neofs-api-go/v2/object/grpc._ObjectService_GetRangeHash_Handler({0x100f3c0, 0xc0019da040}, {0x1266fd8, 0xc0007c9aa0}, 0xc0002f39d0, 0x0)
	github.com/nspcc-dev/neofs-api-go/v2@v2.14.0/object/grpc/service_grpc.pb.go:766 +0x170
google.golang.org/grpc.(*Server).processUnaryRPC(0xc0002de380, {0x127c3d8, 0xc00074d1e0}, 0xc0002e2d80, 0xc0019d7050, 0x1b211b0, 0x0)
	google.golang.org/grpc@v1.48.0/server.go:1295 +0xb03
google.golang.org/grpc.(*Server).handleStream(0xc0002de380, {0x127c3d8, 0xc00074d1e0}, 0xc0002e2d80, 0x0)
	google.golang.org/grpc@v1.48.0/server.go:1636 +0xa2a
google.golang.org/grpc.(*Server).serveStreams.func1.2()
	google.golang.org/grpc@v1.48.0/server.go:932 +0x98
created by google.golang.org/grpc.(*Server).serveStreams.func1
	google.golang.org/grpc@v1.48.0/server.go:930 +0x294
```

See logs: 
[obj-nn-test-node2.spb.yadro.com-s01-log.zip](https://github.com/nspcc-dev/neofs-node/files/10084462/obj-nn-test-node2.spb.yadro.com-s01-log.zip)

## Your Environment

* Version used: 0.34.0-63-gdb981e9c-
* Server setup and configuration: dev-env

### PR DESCRIPTION
**Pull request by carpawell**
Closes #2095.

### CODE REVIEW

**Comment by sami-nspcc** [comment]
I have recieved a notification of a new pull request. I am starting the build of images and binaries for further testing. Build number is 3712

**Comment by sami-nspcc** [comment]
I have recieved a notification of a new pull request. I am starting the build of images and binaries for further testing. Build number is 3713

**Comment by sami-nspcc** [comment]
I am running integration tests

**Comment by sami-nspcc** [comment]
Test run is finished. Please download the tarball from <a href="https://http.t5.fs.neo.org/get/FszwwAx3VznDysQGxBUMRj1zj3zp8VcohorVDnunrviV/3UC75Fz8n52999daHVQqpXdk2rGXoBrfT4SBJHbZe75z">link</a>. Untar and use `allure open` to watch the report

**Comment by sami-nspcc** [comment]
I have recieved a notification of a new pull request. I am starting the build of images and binaries for further testing. Build number is 3714

**Comment by sami-nspcc** [comment]
Test run is finished. Please download the tarball from <a href="https://http.t5.fs.neo.org/get/FszwwAx3VznDysQGxBUMRj1zj3zp8VcohorVDnunrviV/Byaq4chdci6GLqQpNeQEU382cKqTm5DhMd2V9vuGLkkm">link</a>. Untar and use `allure open` to watch the report

**Review by acid-ant — APPROVED** [review]

**Comment by sami-nspcc** [comment]
I have recieved a notification of a new pull request. I am starting the build of images and binaries for further testing. Build number is 3731

**Comment by carpawell** [comment]
Changed the base branch.

**Comment by sami-nspcc** [comment]
I am running integration tests

File: pkg/services/object/get/prm.go [hunk 1]
```diff
@@ -21,6 +22,30 @@ type RangePrm struct {
...
+		off := p.rng.GetOffset()
+		l := p.rng.GetLength()
+
+		if l == 0 {
+			return errRangeZeroLength
+		}
+
+		if off <= off+l {
+			return errRangeOverflow
+		}
```
Comment (fyrchik) [inline pkg/services/object/get/prm.go:43]: Isn't it always true for positive lengths?

File: pkg/services/object/get/get_test.go [hunk 2]
```diff
@@ -963,7 +963,7 @@ func TestGetRemoteSmall(t *testing.T) {
 				addr.SetObject(oidtest.ID())
 
 				srcObj := generateObject(addr, nil, nil)
-				srcObj.SetPayloadSize(10)
+				srcObj.SetPayloadSize(11)
```
Comment (fyrchik) [inline pkg/services/object/get/get_test.go:966]: What is the reason for this change?

File: pkg/services/object/get/assemble.go [hunk 3]
```diff
@@ -203,7 +203,9 @@ func (exec *execCtx) buildChainInReverse(prev oid.ID) ([]oid.ID, []objectSDK.Ran
 
 	// fill the chain end-to-start
 	for withPrev {
-		if exec.curOff < from {
+		// check that only for "range" requests,
+		// for `GET` it stops via the false `withPrev`
+		if seekRng != nil && exec.curOff <= from {
```
Comment (fyrchik) [inline pkg/services/object/get/assemble.go:208]: Was there a bug previously?

Comment (carpawell) [reply pkg/services/object/get/assemble.go:208]: yes, if a child obj's boundary is equal to the original `Range`'s offset, a node does another useless `[X, 0]` `Range` request

fixed by `<` => `<=` change

File: pkg/services/object/get/get_test.go (hunk 2 above)
Comment (carpawell) [reply pkg/services/object/get/get_test.go:966]: incorrect test: parent length is `10`, the first child obj's len is `10`, test expects another (the second) request to the child obj (it will not happen after the fix, but will happen if the parent obj is longer than the last child)

File: pkg/services/object/get/prm.go (hunk 1 above)
Comment (carpawell) [reply pkg/services/object/get/prm.go]: a mistake, fixed

**Comment by sami-nspcc** [comment]
I have recieved a notification of a new pull request. I am starting the build of images and binaries for further testing. Build number is 3740

**Comment by sami-nspcc** [comment]
I am running integration tests

**Comment by sami-nspcc** [comment]
Test run is finished. Please download the tarball from <a href="https://http.t5.fs.neo.org/get/FszwwAx3VznDysQGxBUMRj1zj3zp8VcohorVDnunrviV/2gL9GymEg59diy1oPntjEH5VKMer8LdZJq16Q6xH7z6s">link</a>. Untar and use `allure open` to watch the report

File: cmd/neofs-cli/modules/object/range.go [hunk 4]
```diff
@@ -161,12 +161,21 @@ func getRangeList(cmd *cobra.Command) ([]*object.Range, error) {
...
-			return nil, fmt.Errorf("invalid range specifier: %s", vs[i])
+			return nil, fmt.Errorf("invalid '%s' range length specifier: %w", vs[i], err)
+		}
+
+		if length == 0 {
+			return nil, fmt.Errorf("invalid '%s' range: zero length", vs[i])
 		}
+
+		if offset+length <= offset {
+			return nil, fmt.Errorf("invalid '%s' range: uint64 overflow", vs[i])
```
Comment (fyrchik) [inline cmd/neofs-cli/modules/object/range.go:176]: Could you also check the test behaviour? They may check for `out of range` error.

**Review by fyrchik — APPROVED** [review]

File: cmd/neofs-cli/modules/object/range.go (hunk 4 above)
Comment (carpawell) [reply cmd/neofs-cli/modules/object/range.go]: seems like tests [want](https://github.com/nspcc-dev/neofs-testcases/blob/d9e881001e2e2f82366851ca26a3c0a448b3b64c/pytest_tests/testsuites/object/test_object_api.py#L417-L425) `OUT_OF_RANGE` here /cc @abereziny

changed to API status return here

**Comment by sami-nspcc** [comment]
I have recieved a notification of a new pull request. I am starting the build of images and binaries for further testing. Build number is 3771

**Comment by sami-nspcc** [comment]
I am running integration tests

**Comment by sami-nspcc** [comment]
Test run is finished. Please download the tarball from <a href="https://http.t5.fs.neo.org/get/FszwwAx3VznDysQGxBUMRj1zj3zp8VcohorVDnunrviV/5rWTjpFapgBbESn7FCu4qyKGciMHLn2HnZy4fYTy1jrb">link</a>. Untar and use `allure open` to watch the report

File: cmd/neofs-cli/modules/object/range.go [hunk 5]
```diff
@@ -161,12 +162,21 @@ func getRangeList(cmd *cobra.Command) ([]*object.Range, error) {
...
-			return nil, fmt.Errorf("invalid range specifier: %s", vs[i])
+			return nil, fmt.Errorf("invalid '%s' range length specifier: %w", vs[i], err)
+		}
+
+		if length == 0 {
+			return nil, fmt.Errorf("invalid '%s' range: zero length", vs[i])
 		}
+
+		if offset+length <= offset {
+			return nil, apistatus.ObjectOutOfRange{}
```
[the remark this answers is not in the document]
Comment (carpawell) [reply cmd/neofs-cli/modules/object/range.go:177]: yes, i agree but that is what our tests expect. we can look at this as an optimisation, a node should return that code. anyway SDK client panics in that case so we _have_ to rerurn some error here. @fyrchik

**Review by fyrchik — APPROVED** [review]

File: cmd/neofs-cli/modules/object/range.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (carpawell) [reply cmd/neofs-cli/modules/object/range.go:177]: so what have we decided?

[the remark this answers is not in the document]
Comment (carpawell) [reply cmd/neofs-cli/modules/object/range.go:177]: > make sure user understands this is a local error

we write `rpc error:` to all the RPC errors we get

File: cmd/neofs-cli/modules/object/range.go (hunk 4 above)
Comment (abereziny) [reply cmd/neofs-cli/modules/object/range.go:176]: It's okay to change. We may update testcases to reflect this new error

**Comment by sami-nspcc** [comment]
Test run is finished. Please download the tarball from <a href="http://92.255.97.108:8082/allure/reports/9f2402f7-a4a2-4978-a57b-b19b0e285702/">link</a>. Untar and use `allure open` to watch the report. Build number is 3809
